referendum/referendum.cpp

```diff
@@ -0,0 +1,473 @@
+    check(ref->acts.size(), "ERR::NO_ACTION::No action to execute");
+    check(ref->status == referendum_status::STATUS_PASSING, "ERR:REFERENDUM_NOT_PASSED::Referendum has not passed required number of yes votes");
+    check(calculated_status == referendum_status::STATUS_PASSING, "ERR:REFERENDUM_NOT_PASSED::Referendum has not passed required number of yes votes");
+
+
+    if (ref->type == vote_type::TYPE_BINDING){
+        for (auto a: ref->acts) {
+            a.send();
+        }
+    }
+    else if (ref->type == vote_type::TYPE_SEMI_BINDING){
+        proposeMsig(*ref, dac_id);
+    }
+
+    referenda.erase(ref);
+}
+
+void referendum::stakeobsv(vector<account_stake_delta> stake_deltas, name dac_id){
+    auto dac = dacdir::dac_for_id(dac_id);
+    auto vote_account = dac.account_for_type(dacdir::VOTE_WEIGHT);
+    require_auth(vote_account);
+
+    referenda_table referenda(get_self(), dac_id.value);
+
+    for (auto asd: stake_deltas){
+        votes_table votes(get_self(), dac_id.value);
+        auto existing_vote_data = votes.find(asd.account.value);
+
+        if (existing_vote_data != votes.end()){
+            for (auto v: existing_vote_data->votes){
+                auto ref = referenda.find(v.first);
+
+                if (ref != referenda.end()){
+                    referenda.modify(ref, same_payer, [&](referendum_data& r){
+                        r.token_votes[v.second] += asd.stake_delta.amount;
+                    });
+                }
+            }
+        }
+    }
+}
+
+void referendum::clean(name account, name dac_id){
+    require_auth(account);
+
+    referenda_table referenda(get_self(), dac_id.value);
+    votes_table votes(get_self(), dac_id.value);
+    auto existing_vote_data = votes.find(account.value);
+
+    map<uint64_t, uint8_t> new_votes;
+    if (existing_vote_data != votes.end()){
+        for (auto vd: existing_vote_data->votes){
+            if (referenda.find(vd.first) != referenda.end()){
+                new_votes[vd.first] = vd.second;
+            }
+        }
+    }
+
+    votes.modify(*existing_vote_data, same_payer, [&new_votes](vote_info& v){
+        v.votes = new_votes;
+    });
+}
+
+
+// Private
+
+bool referendum::hasAuth(vector<action> acts){
+    // TODO : this only checks if the permissions provided in the actions can authenticate the action
+    // not if this contract will be able to execute it
+    transaction trx;
+    trx.actions = acts;
+
+    auto check_perms = std::vector<permission_level>();
+    for (auto a: acts){
+        check_perms.insert(check_perms.end(), a.authorization.begin(), a.authorization.end());
+    }
+
+    auto packed_trx = pack(trx);
+    auto packed_perms = pack(check_perms);
+
+    bool res = referendum::_check_transaction_authorization(packed_trx.data(), packed_trx.size(), (const char*)0, 0, packed_perms.data(), packed_perms.size());
+
+    return res;
+}
+
+uint8_t referendum::calculateStatus(uint64_t referendum_id, name dac_id) {
```

Comment:
There seems to be a lot of `ref->voting_type == count_type::COUNT_TOKEN` logic in this function. I think it could be abstracted more to check that condition once near the start of the function to reduce the complexity. Then you could have one `quorum` and `votes` variable used for calculating.

Reply:
Is this just coding style (ie. create intermediate variables), or do you think we should change the underlying config structure?

Reply:
More coding style and maintainability. I think the logic is OK from just reading it.